### ISSUE TITLE
out of disk space - refusing to create the primary instance while the latest generated serial is not zero 

### ISSUE THREAD

**Issue by Tipsmark** [cloudnative-pg/cloudnative-pg#1808]
Kubernetes 1.25 5 
CNPG 1.19

Reproduce: create pgsql instans, run out of disk space = game over.

After data disk runs full the CNPG operator refuses to create a pod again leaving the installation defunct. Resizing a persistent volume cannot take place until the pod is provisioned again.

{"level":"info","ts":"2023-03-24T11:19:44Z","msg":"No primary instance found for this cluster","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"pgsql","namespace":"lala-staging"},"namespace":"lala-staging","name":"pgsql","reconcileID":"c67ecf51-3b9a-4b62-bf5a-c128f5d305d5","uuid":"c70e5a04-ca35-11ed-aad4-46acb41cf6fd"}

Any idea how to fix?

**Comment by sxd** [comment]
@Tipsmark you need to delete the cluster and create it again, this may be a valid issue we will look into this one

**Comment by Tipsmark** [comment]
still an issue - or rather, we had to delete and re-deploy and restore data... but certainly still a problem that a simple but rare case of disk full caused us an unrecoverable instance - happy to hear how others fixed this issue if anyone did.

**Comment by sxd** [comment]
Hi @Tipsmark 

Checking on this, can you add a way to reproduce this issue? also, did you tried to resize the disk before?

Regards!

**Comment by jouve** [comment]
I "reproduced" the same issue today with cloudnative-pg:1.20.1

**Comment by jouve** [comment]
nope, I did not try much and used your methodology (delete/create/restore)

**Comment by Tipsmark** [comment]
And it just happened again - few hundred GB came in, disk full - cluster DEAD... it's becoming a real showstopper that one cannot recover from such a simple problem.

**Comment by Tipsmark** [comment]
Splitting WAL volume to a separate disk didn't help - again database DEAD, gone - ciao... and since backup ALWAYS enables WAL archiving it's rather useless - in our case a simple restore of 500GB db takes 9-12 hours... vs. 15 min. using Kasten, so the network is not the problem... but alas a snapshot based backup doesn't always work when restored... so essentially CNPG is going out the window and being replaced by another solution. It's too fragile and has too few and weak backup options.

**Comment by quercus-carsten** [comment]
Hi Tipsmark,
I am slowly but steadily introducing CNPG in multiple projects. And your case is interesting for me, since you've got way bigger DBs than me. But this is likely to change in the near future.

I understand you are frustrated - and frankly I can relate to that. If I had to restore a DB in production environment, 9-12 hours for a restore would bring me into quite some trouble.
I would rather avoid making the same experience ...

So my question is, would it help to use an autoscaler in your situation?
I found this: [[Kubernetes-Volume-Autoscaler](https://github.com/DevOps-Nirvana/Kubernetes-Volume-Autoscaler)]

From my understanding, you cannot scale down your PVCs afterwards without maintenance - but a sudden spike in disk usage could be cushioned.

And from what I tested, during restore (all?) WALs are beeing processed. There is a lot of random IO involved which is time consuming. Restoring a PVC from a snapshot will always be faster.

There is volume snapshot available since cnpg 1.20.1 [https://github.com/cloudnative-pg/cloudnative-pg/pull/1960](url) , but I didn't look into that yet...

**Comment by Tipsmark** [comment]
@quercus-carsten Not a bad idea, but it will likely not work when using VMWare CSI since we snapshot the FCD's and expansion of a VMDK is not possible when it has a snapshot.  We don't see much IOPS, actually very little, we have 100k+++ available, we see maybe 3k-5k... Even if it was twice or triple the speed it'd still be too slow. We need a 10x improvement to consider it ready for production. Not sure why it can't utilize full bandwidth/cpu/disk... it's just slow with no obvious bottleneck... now we're on 14th hour, restore still running... ingesting wal's at 20-30 Mbit/sec, IOPS 2k, CPU 0.4 of a core, memory a few GB... underlying host near idle, storage near idle - just agonizingly useless.

**Comment by Tipsmark** [comment]
@quercus-carsten so probably last comment there - the restore finally completed, a bit over 60 HOURS! the default config of barman is, at best... useless. I have no intention of trying to tweak it further, instead we've setup workflow to grab the db using pg_dump and the entire dump + upload to S3 takes 46 minutes... a little better than barman's 10 hours and 58 minutes to backup the same database (base backup) - I presume a restore using pg_dump is also going to orders of magnitude faster than restoring the base backup from barman. Whatever the problem is, it needs some serious work before barman can be relied on in a real-world scenario.

**Comment by quercus-carsten** [comment]
Hi @Tipsmark , thanks for sharing. I'm planning to do heavier tests on backup and restore now, too. Some months ago I did prepare workflows and documentation in preparation for the migration of the first projects. But I didn't get to test with databases of dozens of GBs in size - let alone hundreds...

I have zero experience with MSSQL. But roughly 600GB of data in approx 15 min - I mean this can't involve applying WALs or "redo logs" or whatever this may be called in this case, or can it? I mean does this also include PITR? I can only image that MS does this snapshot based without any catching up to latest timeline.

Anyway, I find this to be extremely fast  - and from the metrics you gave and the underlying infrastructure I guess one has to look closer on optimizing barman cloud restore.
I first stumbled upon bandwidth limitation settings regarding barman - but this does not apply to barman-cloud-backup and barman-cloud-restore / barman-wal-restore. The only setting I found which I will test is "--parallel". Defaults to 2 if I'm not mistaken.

Also interesting to hear, that pg_dump works this fast... But PITR is essential for me, so I really am dependent on it.
As soon as I did my own tests, I will share the results here.

**Comment by gbartolini** [comment]
> still an issue - or rather, we had to delete and re-deploy and restore data... but certainly still a problem that a simple but rare case of disk full caused us an unrecoverable instance - happy to hear how others fixed this issue if anyone did.

Can you please help us understand why you ran out of disk? Was it because of a spike in the WAL production, WAL archiving to be lagging, or increase in the database size?

**Comment by divyangjp** [comment]
It's still happening in v1.22.1
I used pgbench to add dummy data beyond the capacity of underlying storage. 
It made the whole setup and wal archives useless. Couldn't recover into new instance either.

**Comment by bonesmoses** [comment]
I find this particular issue somewhat odd. It is not standard for Postgres to become irrecoverable when disk space is exceeded. At worst, in-flight transactions cannot be written and the database halts. Once disk storage is increased, it should be able to resume following a short recovery phase. is there something the operator itself is doing to prevent this from occurring?

**Comment by Routhinator** [comment]
This sounds like a deficiency in the CSI you are using. I hit this problem this week, adjusted the volume size, and my CSI driver (democratic-csi) did the resize of the volume and the partition on the volume, and I was back up and running.

If your CSI driver doesn't resize the partition (hence the partition needs to be online part) then:

- Apply the manual intevention step from https://cloudnative-pg.io/documentation/current/failure_modes/#failure-modes_1 to stop the self healing loop.
- Delete any db instance pods that are deployed
- Craft a Pod deployment that mounts the volumes of the DBs using a container image that contains resize2fs as a cluster admin so you have the appropriate rights, the container will need to run as root.
- Use resizesf2 to scale the partitions to the newly adjusted volume size
- Delete your crafted pod to release the volumes
- Remove the manual intervention property to allow the self-healing loop to resume
- CNPG will take back over and start the db pods, and you'll be able to recover.

**Comment by Routhinator** [comment]
So I just hit this again today and this time it's just bonkers...

* WAL file backups and archives are working and all on S3, however db-1 thought it was failing to upload the history file for some reason (it wasn't - those were there too) and as a result it was not deleting WALs.
* Of course, this lead to a full disk
* For some reason, even though all the WALs were on S3, db-2 was not applying them, and now it's primary and broken.

So from this state - I am trying to recover. I deleted the already backed up WALs from db-1's PVC, and I am struggling to get CNPG to force it back to primary, because it keeps trying to copy the garbage data from db-2 which I killed with fire since it's state was inconsistent and it should never have been promoted in the first place... 

And I'm going to have to likely wipe and restore from backup to get out of this..

**Comment by passcod** [comment]
I also have this issue. 100MB of actual data, 10GB of disk space, and CNPG randomly stops deleting WALs even though they appear to be in S3, and completely takes down a service. Insane. This makes CNPG completely unsuitable for any production usage, as the only "solution" is to not have backups enabled.

**Comment by Routhinator** [comment]
Not stale, not solved.

**Comment by quercus-carsten** [comment]
Not solved. Bump.

**Comment by Routhinator** [comment]
Not stale

**Comment by Routhinator** [comment]
Not stale

**Comment by Routhinator** [comment]
Yeah this needs to be addressed still. Its a major edge in this operator. @sxd

**Comment by sxd** [comment]
@Routhinator please open a new issue with a description and a way to reproduce it, since this, a lot of things has changed
